src/bookmarks.rs

```diff
@@ -0,0 +1,95 @@
+use crate::config::Flattenable;
+use crate::error::JoshutoResult;
+use serde_derive::Deserialize;
+use serde_derive::Serialize;
+use std::collections::HashMap;
+use std::io::Write;
+use std::path::PathBuf;
+#[derive(Debug, Clone, Deserialize, Serialize)]
+pub struct RawBookmarkMapping {
+    #[serde(default)]
+    pub bookmarks: Vec<(char, String)>,
+}
+
+impl RawBookmarkMapping {
+    fn save(&self, path: &str) -> std::io::Result<()> {
+        let s = toml::to_string_pretty(self);
+        if let Ok(s) = s {
+            let mut file = std::fs::OpenOptions::new()
+                .write(true)
+                .create(true)
+                .open(path)
+                .unwrap();
+
+            file.write_all(&s.as_bytes()).unwrap();
+        }
+        Ok(())
+    }
+}
+
+impl Flattenable<BookmarkMapping> for RawBookmarkMapping {
+    fn flatten(self) -> BookmarkMapping {
+        let mut bookmarksmap = BookmarkMapping::new();
+        for (c, p) in self.bookmarks {
+            bookmarksmap.map.insert(c, PathBuf::from(p));
+        }
+        // notify(&bookmarks);
+        bookmarksmap
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct BookmarkMapping {
+    pub map: HashMap<char, PathBuf>,
+}
+
+impl BookmarkMapping {
+    pub fn save(&self, path: &str) -> std::io::Result<()> {
+        self.to_raw().save(path)?;
+        Ok(())
+    }
+
+    pub fn to_raw(&self) -> RawBookmarkMapping {
+        let mut bookmarks = vec![];
+        for (k, v) in self.map.iter() {
+            bookmarks.push((*k, v.to_str().unwrap().to_string()));
+        }
+        return RawBookmarkMapping { bookmarks };
+    }
+
+    pub fn new() -> Self {
+        let map = HashMap::new();
+        Self { map }
+    }
+
+    pub fn load(bookmarks_file_path: &str) -> Self {
```

Comment:
This could return `std::io::Result<Self>`, `Result<Self, String>` or something and delegate the error handling to elsewhere.